### PR TITLE
NSString(contentsOfFile:usedEncoding:) fixes.

### PR DESCRIPTION
**Pull request by dplanitzer**
- passing nil to the usedEncoding: parameter no longer causes the NSString(contentsOf:usedEncoding:) to terminate with an error.
- implemented NSString(contentsOfFile:usedEncoding:)
- String(contentsOfFile:) now works too

### CODE REVIEW

**Comment by spevans** [comment]
@swift-ci please test

File: Foundation/NSString.swift [hunk 1]
```diff
@@ -1248,33 +1248,35 @@ extension NSString {
...
+            encoding = String.Encoding.utf8.rawValue
+            offset = 0
         }
 
         // Since the encoding being passed includes the byte order the BOM wont be checked or skipped, so pass offset to
         // manually skip the BOM header.
-        guard let enc = enc, let cf = CFStringCreateWithBytes(kCFAllocatorDefault, bytePtr + offset, readResult.length - offset,
-                                                              CFStringConvertNSStringEncodingToEncoding(enc.pointee), true) else {
+        guard let cf = CFStringCreateWithBytes(kCFAllocatorDefault, bytePtr + offset, readResult.length - offset,
+                                                              CFStringConvertNSStringEncodingToEncoding(encoding), true) else {
```
Comment (xwu) [inline Foundation/NSString.swift]: Drive-by nit: Please align this line with the parenthesis of the previous line, as was originally done.

File: Foundation/NSString.swift [hunk 2]
```diff
@@ -1286,11 +1288,13 @@ extension NSString {
             throw NSError(domain: NSCocoaErrorDomain, code: CocoaError.fileReadInapplicableStringEncoding.rawValue, userInfo: [
                 "NSDebugDescription" : "Unable to bridge CFString to String."
                 ])
-        }    
+        }
+        
+        enc?.pointee = encoding
```
Comment (ianpartridge) [inline Foundation/NSString.swift]: Previously the code set `enc?.pointee` even if the `CFStringCreateWithBytes()` called failed and the function threw.  Now it doesn't.  Did you check what the behaviour of Darwin is in this regard?

**Comment by dplanitzer** [comment]
@xwu: formatting is fixed now

@ianpartridge: I didn't cross check with the Darwin implementation. However, I've changed the code such that we now set the encoding out parameter before we call the CFStringCreate() function.

**Comment by phausler** [comment]
this is on-par with the way the objc version does it

@swift-ci please test and merge

**Comment by alblue** [comment]
@swift-ci please test and merge

**Comment by alblue** [comment]
Transient failure for the GitHub API requesting merge - the build succeeded.

**Comment by alblue** [comment]
@swift-ci please test and merge
